controllers/tinkerbellcluster_controller.go

```diff
@@ -163,42 +161,63 @@ var (
 	ErrControlPlaneEndpointNotSet = fmt.Errorf("controlplane endpoint is not set")
 )
 
-func nextAvailableHardware(ctx context.Context, k8sClient client.Client, extraSelectors []string) (*tinkv1.Hardware, error) { //nolint:lll
-	hardware, err := nextHardware(ctx, k8sClient, append(extraSelectors, fmt.Sprintf("!%s", HardwareOwnerNameLabel)))
-	if err != nil {
-		return nil, fmt.Errorf("getting next Hardware object: %w", err)
-	}
-
-	if hardware == nil {
-		return nil, ErrNoHardwareAvailable
-	}
-
-	return hardware, nil
-}
-
-func nextHardware(ctx context.Context, k8sClient client.Client, selectors []string) (*tinkv1.Hardware, error) { //nolint:lll
+func nextHardware(ctx context.Context, k8sClient client.Client, selectors *metav1.LabelSelector) (*tinkv1.Hardware, error) { //nolint:lll
 	availableHardwares := &tinkv1.HardwareList{}
+	for {
+		selector, err := metav1.LabelSelectorAsSelector(selectors)
+		if err != nil {
+			return nil, fmt.Errorf("converting label selector: %w", err)
+		}
+		if err := k8sClient.List(ctx, availableHardwares, &client.ListOptions{LabelSelector: selector}); err != nil {
+			return nil, fmt.Errorf("listing hardware without owner: %w", err)
+		}
 
-	selectorsRaw := strings.Join(selectors, ",")
+		// found some hardware
+		if len(availableHardwares.Items) > 0 {
+			return &availableHardwares.Items[0], nil
+		}
 
-	selector, err := labels.Parse(selectorsRaw)
-	if err != nil {
-		return nil, fmt.Errorf("parsing raw labels selector %q: %w", selectorsRaw, err)
+		// couldn't find anything, so try to be relax our search
+		if !relaxHardwareLabelSelector(selectors) {
+			break
+		}
 	}
 
-	options := client.MatchingLabelsSelector{
-		Selector: selector,
-	}
+	// nothing was found and we can't relax our label selector any more
+	return nil, ErrNoHardwareAvailable
+}
 
-	if err := k8sClient.List(ctx, availableHardwares, options); err != nil {
-		return nil, fmt.Errorf("listing hardware without owner: %w", err)
+// relaxHardwareLabelSelector attempts to remove requirements from the selectors sequentially to broaden the search.  If
+// no more can be removed, the function returns false.
+func relaxHardwareLabelSelector(selectors *metav1.LabelSelector) bool {
+	// special case to first drop the owner labels
+	if _, ok := selectors.MatchLabels[HardwareOwnerNameLabel]; ok {
+		delete(selectors.MatchLabels, HardwareOwnerNameLabel)
+		delete(selectors.MatchLabels, HardwareOwnerNamespaceLabel)
+		selectors.MatchExpressions = append(selectors.MatchExpressions, metav1.LabelSelectorRequirement{
+			Key:      HardwareOwnerNameLabel,
+			Operator: metav1.LabelSelectorOpDoesNotExist,
+		})
+		return true
+	}
+
+	// TODO: is this necessary or desired? It provides fallback in the event you request hardware in rack=x, but there
```

Comment:
I'm specifically not sure about this, the idea was that the machine label selectors effectively acted as preferences rather than required requirements.

Reply:
I think if we're doing just basic label selection, we want it to be more strict. As-is, this would result in choosing hardware that the cluster creator might not want. Does an `affinity` field similar to pods make sense on the `TinkerbellMachineSpec` if we wanted to support optional and required?

```go

type HardwareAffinityTerm struct {
    Selector *metav1.LabelSelector `json:"selector,omitempty"`
    Weight   int32                 `json:"weight,omitempty"`
}

type HardwareAffinity {
    Preferred []HardwareAffinityTerm `json:"preferred,omitempty"`
    Required  []HardwareAffinityTerm `json:"required,omitempty"`
}

type TinkerbellMachineSpec struct {
    //...
    HardwareAffinity HardwareAffinity `json:"hardwareAffinity,omitempty"`
}
```

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: TinkerbellMachineTemplate
metadata:
  name: something
  namespace: somewhere
spec:
  template:
    spec:
      hardwareAffinity:
        required:
        - weight: 15
          selector:
            matchLabels: 
              manufacturer: dell
            matchExpressions:
            - key: idracVersion
              operator: In
              values: ["8", "9"]
        preferred:
        - weight: 50
          selector:
            matchLabels: 
              rack: 1
              room: 2
        
```

Reply:
the `Preferred` might be overkill for an initial pass, but if we just did `Required`, then it would essentially be what you have here, just shoved down into a substructure

Reply:
SGTM